pkg/reconcilers/base_reconciler.go

```diff
@@ -0,0 +1,108 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/3scale/apicast-operator/pkg/k8sutils"
+)
+
+// MutateFn is a function which mutates the existing object into it's desired state.
+type MutateFn func(existing, desired k8sutils.KubernetesObject) (bool, error)
+
+func CreateOnlyMutator(existing, desired k8sutils.KubernetesObject) (bool, error) {
+	return false, nil
+}
+
+type BaseReconciler struct {
+	// client should be a split client that reads objects from
+	// the cache and writes to the Kubernetes APIServer
+	client client.Client
+	// apiClientReader should be a client that directly reads objects
+	// from the Kubernetes APIServer
+	apiClientReader client.Reader
+	scheme          *runtime.Scheme
+	logger          logr.Logger
+}
+
+func NewBaseReconciler(client client.Client, apiClientReader client.Reader, scheme *runtime.Scheme, logger logr.Logger) BaseReconciler {
+	return BaseReconciler{
+		client:          client,
+		apiClientReader: apiClientReader,
+		scheme:          scheme,
+		logger:          logger,
+	}
+}
+
+func (b *BaseReconciler) Client() client.Client {
+	return b.client
+}
+
+func (b *BaseReconciler) APIClientReader() client.Reader {
+	return b.apiClientReader
+}
+
+func (b *BaseReconciler) Scheme() *runtime.Scheme {
+	return b.scheme
+}
+
+func (b *BaseReconciler) Logger() logr.Logger {
+	return b.logger
+}
+
+func (b *BaseReconciler) ReconcileResource(existing, desired k8sutils.KubernetesObject, mutateFn MutateFn) error {
```

Comment:
Can we add some comment to explain what is the expected value of `existing`  and what's the behaviour in relation to it?

It would help clarify the behaviour of the method for users of it

Reply:
:+1:

Reply:
Some rewording done. Review required @miguelsorianod

Reply:
:+1: